Source/WebCore/dom/InternalObserverToArray.cpp

```diff
@@ -0,0 +1,102 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "InternalObserverToArray.h"
+
+#include "Exception.h"
+#include "ExceptionCode.h"
+#include "InternalObserver.h"
+#include "JSDOMPromiseDeferred.h"
+#include "Observable.h"
+#include "ScriptExecutionContext.h"
+#include "SubscribeOptions.h"
+#include "Subscriber.h"
+#include "SubscriberCallback.h"
+#include <JavaScriptCore/JSCJSValueInlines.h>
+#include <wtf/Vector.h>
+
+namespace WebCore {
+
+class InternalObserverToArray final : public InternalObserver {
+public:
+    static Ref<InternalObserverToArray> create(ScriptExecutionContext& context, Ref<DeferredPromise>&& promise)
+    {
+        Ref internalObserver = adoptRef(*new InternalObserverToArray(context, WTFMove(promise)));
+        internalObserver->suspendIfNeeded();
+        return internalObserver;
+    }
+
+private:
+    void next(JSC::JSValue value) final
+    {
+        auto* globalObject = protectedScriptExecutionContext()->globalObject();
+        ASSERT(globalObject);
+        m_list.append({ globalObject->vm(), value });
+    }
+
+    void error(JSC::JSValue value) final
+    {
+        protectedPromise()->reject<IDLAny>(value);
+    }
+
+    void complete() final
+    {
+        InternalObserver::complete();
+        protectedPromise()->resolve<IDLSequence<IDLAny>>(WTFMove(m_list));
+    }
+
+    void visitAdditionalChildren(JSC::AbstractSlotVisitor&) const final
+    {
+    }
+
+    Ref<DeferredPromise> protectedPromise() const { return m_promise; }
+
+    InternalObserverToArray(ScriptExecutionContext& context, Ref<DeferredPromise>&& promise)
+        : InternalObserver(context)
+        , m_promise(WTFMove(promise))
+    {
+    }
+
+    Vector<JSC::Strong<JSC::Unknown>> m_list;
```

Comment:
Personally, I'd feel better if we didn't use JSC::Strong for data members. It is too easy to cause leaks. Could we use JSValueInWrappedObject inside the Vector and visit the values inside the `visitAdditionalChildren(JSC::AbstractSlotVisitor&)` member function in this class? We can convert to a Vector of JSC::Strong in `complete()`

Reply:
Would we loop every member of the vector in visitAdditonalChildren? And if so are we concerned with say a vector with thousands of members?

Would be nice if there was something like a VectorWrapped

Reply:
I don't know if there is a more efficient data type. @MenloDorian or @Constellation may have a proposal.